### DISCUSSION TITLE
How to check that files have not been modified.

### DISCUSSION THREAD

**Discussion by ghost** [ArchiveBox/ArchiveBox#1533] [Q&A] [answered]
Is there a way to prove that the stored files were not modified?

For example, if I add a direct link to a `.txt`, after downloading it I can go into the `data` directory/volume and modify the contents of the file.

I would like to understand how to "prove" that the file was not altered.

**Comment by pirate** [comment]
I have some code that generates a merkle tree sha256 and blake3 of each result. It wouldn't be too hard to submit it to an external audit log and check hashes periodically.

We also need to store the attestation signatures of the HTTPS content using something like TLSNotary. 

For now I recommend solving this at the filesystem layer with ZFS. It lets you take incremental immutable snapshots and checksum verify them.

**Reply by ghost** [reply]
It hadn't occurred to me to use a file system with snapshots.

Now that I think about it, the question would be whether some kind of 'hook' is provided to act when the snapshot (of the website/url) is complete and then decide what to do.

**Reply by pirate** [reply, marked as answer]
That's how the new plugin system is implemented. Take a look at the newmodels branch open PR, and look inside archivebox/abx/archivebox/events.py and the other files in that dir. 

We provide hooks for pre/post save, delete update, etc

**Reply by ghost** [reply]
Yes, I just saw that it is in the 'dev' branch, it was in the stable branch and I didn't see the hooks option.
